swupd/manifest.go

```diff
@@ -319,6 +319,32 @@ func (m *Manifest) WriteManifest(w io.Writer) error {
 	return nil
 }
 
+// filterFilesVersion returns a new manifest with only Files and DeletedFiles
+// updated in versions greater than fromVersion
+func (m *Manifest) filterFilesVersion(fromVersion uint32) *Manifest {
+	fm := &Manifest{
+		Header:     m.Header,
+		Name:       m.Name,
+		BundleInfo: m.BundleInfo,
+	}
+	fm.Header.ContentSize = 0
+
+	for _, f := range m.Files {
```

Comment:
This function could be a lot simpler if you just sort the files by version then search for the first instance of the version and return files[firstInstance:]

Reply:
This function is O(n) and sorting is O(n*log(n)) or worse. It's preferable to sort the list after filtering so we're going to sort a smaller list.

The ideal would be to use something similar to python filters. Is there something like that in GoLang?

    fm = [f if f.Version > fromVersion]

or

    fm = filter(lambda x: x.Version > fromVersion, f)

Reply:
You can use lambdas and simulate filters in Go, but it's really not efficient vs a regular for loop because it does a ton of extra work & copying:
https://github.com/robpike/filter
tl;dr rob pike also created Go and said don't use it. It will likely be much more doable once Go supports generics and this can be implemented more efficiently.

Reply:
thanks. I prefer to let this function the way it is. If you prefer sorting + selecting for readability I can change that.

Reply:
No I don't prefer sorting ahead of time, this looks ok. I think obsessively sorting is something we need to look at again in both mixer & swupd, 'cause I think there are some places where we do it too much "just in case."

Reply:
There are probably cases where we do it too much just in case, but I think instead of removing operations on sorted slices we should consider 1) inserting elements in the correct location to generate a sorted list or 2) convert to using a set instead of a slice for this sort of thing. We should *never* have implemented these as slices.

Reply:
1) I have a strong feeling that keeping the list sorted is worse than sorting later.
2) Assuming a Set is a hash table we're going to be faster on searches if we use that, but adding and sorting is going to be the same as using a list. I don't know how this structure is used in other areas of mixer to know if it's worth changing that.

Reply:
I think I understood your point on (1) now. The point isn't just about replacing a sort by insertion on a sorted list. it is about not trying to sort it more than once. It makes sense in this case. But maybe (2) is even better if there are searches.

Reply:
I'm not sure about the usefulness of (2) in regards to searches. IIRC there are only a couple of times we need to search for something without iterating over the entire set anyways.

For many operations a sorted list is very important, such as when we `linkPeersAndChange`. We walk like that a couple times in the code so we should (at some point, not now) evaluate if it is better to sort once after list construction or if we should always insert in the correct place.

Part of that evaluation has to check if we create the slice, operate on the sorted slice, then insert again and operate again. If this is the case it makes sense to do correctly sorted inserts, at least for the second and further iterations.

Reply:
If you need to sort often (i.e after every few inserts or deletes), then using a balanced tree to insert/remove quickly and keep things sorted as elements are added/removed is way better.
If we do many inserts/deletes and only require 1 sort at the end, then we'll get better performance just using a regular slice to add/remove unsorted and then call sort once when we're done.